File: app/src/main/java/remix/myplayer/ui/adapter/AlbumAdapter.java
package remix.myplayer.ui.adapter;

import static remix.myplayer.request.ImageUriRequest.BIG_IMAGE_SIZE;
import static remix.myplayer.request.ImageUriRequest.SMALL_IMAGE_SIZE;

import android.annotation.SuppressLint;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import com.facebook.drawee.view.SimpleDraweeView;
import com.github.promeg.pinyinhelper.Pinyin;
import remix.myplayer.App;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Album;
import remix.myplayer.databinding.ItemAlbumRecycleGridBinding;
import remix.myplayer.databinding.ItemAlbumRecycleListBinding;
import remix.myplayer.misc.menu.LibraryListener;
import remix.myplayer.theme.Theme;
import remix.myplayer.theme.ThemeStore;
import remix.myplayer.ui.adapter.holder.BaseViewHolder;
import remix.myplayer.ui.adapter.holder.HeaderHolder;
import remix.myplayer.ui.misc.MultipleChoice;
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller;
import remix.myplayer.util.Constants;
import remix.myplayer.util.ImageUriUtil;
import remix.myplayer.util.ToastUtil;

/**
 * Created by Remix on 2015/12/20.
 */

/**
 * 专辑界面的适配器
 */
public class AlbumAdapter extends HeaderAdapter<Album, BaseViewHolder> implements
    FastScroller.SectionIndexer {

  public AlbumAdapter(int layoutId, MultipleChoice multipleChoice,
      RecyclerView recyclerView) {
    super(layoutId, multipleChoice, recyclerView);
  }

  @NonNull
  @Override
  public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    if (viewType == TYPE_HEADER) {
      return new HeaderHolder(
          LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_header_2, parent, false));
    }
    return viewType == HeaderAdapter.LIST_MODE ?
        new AlbumListHolder(ItemAlbumRecycleListBinding
            .inflate(LayoutInflater.from(parent.getContext()), parent, false)) :
        new AlbumGridHolder(ItemAlbumRecycleGridBinding
            .inflate(LayoutInflater.from(parent.getContext()), parent, false));
  }

  @Override
  public void onViewRecycled(@NonNull BaseViewHolder holder) {
    super.onViewRecycled(holder);
    disposeLoad(holder);
  }

  @SuppressLint("RestrictedApi")
  @Override
  protected void convert(BaseViewHolder baseHolder, Album album, int position) {
    if (position == 0) {
      final HeaderHolder headerHolder = (HeaderHolder) baseHolder;
      setUpModeButton(headerHolder);
      return;
    }

    if (!(baseHolder instanceof AlbumHolder)) {
      return;
    }

    final Context context = baseHolder.itemView.getContext();
    final AlbumHolder holder = (AlbumHolder) baseHolder;
    holder.mText1.setText(album.getAlbum());

    //设置封面
    final long albumId = album.getAlbumID();
    final int imageSize = mMode == LIST_MODE ? SMALL_IMAGE_SIZE : BIG_IMAGE_SIZE;
    holder.mImage.setTag(setImage(holder.mImage, ImageUriUtil.getSearchRequest(album), imageSize, position));

    if (holder instanceof AlbumListHolder) {
      holder.mText2
          .setText(App.getContext().getString(R.string.song_count_2, album.getArtist(), album.getCount()));
    } else {
      holder.mText2.setText(album.getArtist());
    }

    holder.mContainer.setOnClickListener(v -> {
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg);
        return;
      }
      mOnItemClickListener.onItemClick(holder.mContainer, position - 1);
    });
    //多选菜单
    holder.mContainer.setOnLongClickListener(v -> {
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg);
        return true;
      }
      mOnItemClickListener.onItemLongClick(holder.mContainer, position - 1);
      return true;
    });

    //着色
    int tintColor = ThemeStore.getLibraryBtnColor();
    Theme.tintDrawable(holder.mButton, R.drawable.icon_player_more, tintColor);

    holder.mButton.setOnClickListener(v -> {
      if (mChoice.isActive()) {
        return;
      }
      final PopupMenu popupMenu = new PopupMenu(context, holder.mButton, Gravity.END);
      popupMenu.getMenuInflater().inflate(R.menu.menu_album_item, popupMenu.getMenu());
      popupMenu.setOnMenuItemClickListener(new LibraryListener(
          context,
          albumId,
          Constants.ALBUM,
          album.getAlbum()));
      popupMenu.show();
    });

    //是否处于选中状态
    holder.mContainer.setSelected(mChoice.isPositionCheck(position - 1));

    //半圆着色
    if (mMode == HeaderAdapter.GRID_MODE) {
      Theme.tintDrawable(holder.mHalfCircle, R.drawable.icon_half_circular_left,
          ThemeStore.getBackgroundColorMain(context));
    }

    setMarginForGridLayout(holder, position);
  }


  @Override
  public String getSectionText(int position) {
    if (position == 0) {
      return "";
    }
    if (mDatas != null && position - 1 < mDatas.size()) {
      String album = mDatas.get(position - 1).getAlbum();
      return !TextUtils.isEmpty(album) ? (Pinyin.toPinyin(album.charAt(0))).toUpperCase()
          .substring(0, 1) : "";
    }
    return "";
  }

  static class AlbumHolder extends BaseViewHolder {

    @Nullable
    ImageView mHalfCircle;
    TextView mText1;
    TextView mText2;
    ImageButton mButton;
    SimpleDraweeView mImage;
    ViewGroup mContainer;

    AlbumHolder(View v) {
      super(v);
    }
  }

  static class AlbumGridHolder extends AlbumHolder {

    AlbumGridHolder(ItemAlbumRecycleGridBinding binding) {
      super(binding.getRoot());
      mHalfCircle = binding.itemHalfCircle;
      mText1 = binding.itemText1;
      mText2 = binding.itemText2;
      mButton = binding.itemButton;
      mImage = binding.itemSimpleiview;
      mContainer = binding.itemContainer;
    }
  }

  static class AlbumListHolder extends AlbumHolder {

    AlbumListHolder(ItemAlbumRecycleListBinding binding) {
      super(binding.getRoot());
      mText1 = binding.itemText1;
      mText2 = binding.itemText2;
      mButton = binding.itemButton;
      mImage = binding.itemSimpleiview;
      mContainer = binding.itemContainer;
    }
  }

}

File: app/src/main/java/remix/myplayer/ui/adapter/AlbumAdapter.kt
package remix.myplayer.ui.adapter

import android.annotation.SuppressLint
import android.graphics.Color
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.MultiTransformation
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.signature.ObjectKey
import com.github.promeg.pinyinhelper.Pinyin
import remix.myplayer.App
import remix.myplayer.R
import remix.myplayer.bean.mp3.Album
import remix.myplayer.databinding.ItemAlbumRecycleGridBinding
import remix.myplayer.databinding.ItemAlbumRecycleListBinding
import remix.myplayer.glide.UriFetcher
import remix.myplayer.helper.SortOrder
import remix.myplayer.misc.menu.LibraryListener
import remix.myplayer.theme.Theme
import remix.myplayer.theme.ThemeStore.getBackgroundColorMain
import remix.myplayer.theme.ThemeStore.libraryBtnColor
import remix.myplayer.ui.adapter.holder.BaseViewHolder
import remix.myplayer.ui.adapter.holder.HeaderHolder
import remix.myplayer.ui.misc.MultipleChoice
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller
import remix.myplayer.util.Constants
import remix.myplayer.util.DensityUtil
import remix.myplayer.util.SPUtil
import remix.myplayer.util.SPUtil.SETTING_KEY
import remix.myplayer.util.ToastUtil
import java.lang.ref.WeakReference
import java.util.*


/**
 * Created by Remix on 2015/12/20.
 */
/**
 * 专辑界面的适配器
 */
class AlbumAdapter(layoutId: Int, multipleChoice: MultipleChoice<Album>,
                   recyclerView: RecyclerView) : HeaderAdapter<Album, BaseViewHolder>(layoutId, multipleChoice, WeakReference(recyclerView)), FastScroller.SectionIndexer {

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
    if (viewType == TYPE_HEADER) {
      return HeaderHolder(
          LayoutInflater.from(parent.context).inflate(R.layout.layout_header_2, parent, false))
    }
    return if (viewType == LIST_MODE) AlbumListHolder(ItemAlbumRecycleListBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    else AlbumGridHolder(ItemAlbumRecycleGridBinding.inflate(LayoutInflater.from(parent.context), parent, false))
  }


  @SuppressLint("RestrictedApi")
  override fun convert(holder: BaseViewHolder, album: Album?, position: Int) {
    if (position == 0) {
      val headerHolder = holder as HeaderHolder
      setUpModeButton(headerHolder)
      return
    }
    if (holder !is AlbumHolder || album == null) {
      return
    }
    val context = holder.itemView.context
    holder.tv1.text = album.album

    //设置封面
    val options = RequestOptions()
        .placeholder(Theme.resolveDrawable(holder.itemView.context, R.attr.default_album))
        .error(Theme.resolveDrawable(holder.itemView.context, R.attr.default_album))

    if (mode == GRID_MODE) {
      options.transform(MultiTransformation(CenterCrop(), RoundedCorners(DensityUtil.dip2px(2f))))
    }

    Glide.with(holder.itemView)
        .load(album)
        .apply(options)
        .signature(ObjectKey(UriFetcher.albumVersion))
        .into(holder.iv)

    if (holder is AlbumListHolder) {
      holder.tv2.text = App.context.getString(R.string.song_count_2, album.artist, album.count)
    } else {
      holder.tv2.text = album.artist
    }
    holder.container.setOnClickListener { v: View? ->
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg)
        return@setOnClickListener
      }
      onItemClickListener?.onItemClick(holder.container, position - 1)
    }
    //多选菜单
    holder.container.setOnLongClickListener { v: View? ->
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg)
        return@setOnLongClickListener true
      }
      onItemClickListener?.onItemLongClick(holder.container, position - 1)
      true
    }

    //着色
    val tintColor = libraryBtnColor
    Theme.tintDrawable(holder.btn, R.drawable.icon_player_more, tintColor)
    holder.btn.setOnClickListener { v: View? ->
      if (choice.isActive) {
        return@setOnClickListener
      }
      val popupMenu = PopupMenu(context, holder.btn, Gravity.END)
      popupMenu.menuInflater.inflate(R.menu.menu_album_item, popupMenu.menu)
      popupMenu.setOnMenuItemClickListener(LibraryListener(
          context, album,
          Constants.ALBUM,
          album.album))
      popupMenu.show()
    }

    //是否处于选中状态
    holder.container.isSelected = choice.isPositionCheck(position - 1)

    //半圆着色
    if (mode == GRID_MODE) {
      Theme.tintDrawable(holder.ivHalfCircle, R.drawable.icon_half_circular_left,
        Color.RED)
    }
    setMarginForGridLayout(holder, position)
  }

  override fun getSectionText(position: Int): String {
    if (position in 1..dataList.size) {
      val data = dataList[position - 1]
      val key = when (SPUtil.getValue(
          App.context,
          SETTING_KEY.NAME,
          SETTING_KEY.ALBUM_SORT_ORDER,
          SortOrder.ALBUM_A_Z
      )) {
        SortOrder.ALBUM_A_Z, SortOrder.ALBUM_Z_A -> data.album
        SortOrder.ARTIST_A_Z, SortOrder.ARTIST_Z_A -> data.artist
        else -> ""
      }
      if (key.isNotEmpty())
        return Pinyin.toPinyin(key[0]).uppercase(Locale.getDefault()).substring(0, 1)
    }
    return ""
  }

  internal open class AlbumHolder(v: View?) : BaseViewHolder(v) {
    lateinit var ivHalfCircle: ImageView
    lateinit var tv1: TextView
    lateinit var tv2: TextView
    lateinit var btn: ImageButton
    lateinit var iv: ImageView
    lateinit var container: ViewGroup
  }

  internal class AlbumGridHolder(binding: ItemAlbumRecycleGridBinding) : AlbumHolder(binding.root) {
    init {
      ivHalfCircle = binding.itemHalfCircle
      tv1 = binding.itemText1
      tv2 = binding.itemText2
      btn = binding.itemButton
      iv = binding.iv
      container = binding.itemContainer
    }
  }

  internal class AlbumListHolder(binding: ItemAlbumRecycleListBinding) : AlbumHolder(binding.root) {
    init {
      tv1 = binding.itemText1
      tv2 = binding.itemText2
      btn = binding.itemButton
      iv = binding.iv
      container = binding.itemContainer
    }
  }
}